ledger/ledger-on-posix-filesystem/src/main/scala/com/daml/ledger/on/filesystem/posix/FileSystemLedgerReaderWriter.scala

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.on.filesystem.posix
+
+import java.nio.file.{Files, NoSuchFileException, Path}
+import java.time.Clock
+import java.util.UUID
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.on.filesystem.posix.FileSystemLedgerReaderWriter._
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlLogEntryId,
+  DamlStateKey,
+  DamlStateValue,
+  DamlSubmission
+}
+import com.daml.ledger.participant.state.kvutils.api.{LedgerReader, LedgerRecord, LedgerWriter}
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueCommitting}
+import com.daml.ledger.participant.state.v1.{LedgerId, Offset, ParticipantId, SubmissionResult}
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+import com.digitalasset.platform.akkastreams.dispatcher.Dispatcher
+import com.digitalasset.platform.akkastreams.dispatcher.SubSource.OneAfterAnother
+import com.google.protobuf.ByteString
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Random, Success}
+
+class FileSystemLedgerReaderWriter private (
```

Comment:
It would be great to have a quick recap of the design here as a comment. I.e., what do we store on the different paths and how are they connected.

Reply:
I have added comments for all the different paths.